### PR TITLE
Add pagination support to ReferencesClient.

### LINKED ISSUE #1664: Add missing pagination support to References Clients
Refer #1659 

## Implement Pagination on the following clients

[IReferencesClient](https://github.com/octokit/octokit.net/blob/master/Octokit/Clients/IReferencesClient.cs) and [IObservableReferencesClient](https://github.com/octokit/octokit.net/blob/master/Octokit.Reactive/Clients/IObservableReferencesClient.cs)
- `GetAll(string owner, string name)`
- `GetAll(long repositoryId)`
- `GetAllForSubNamespace(string owner, string name, string subNamespace)`
- `GetAllForSubNamespace(long repositoryId, string subNamespace)`

## Checklist

- [ ] Remove convention test exclusion

- [ ] Add `ApiOptions` overload
  - make existing methods call new overload with `ApiOptions.None`

- [ ] Update unit tests
  - Adjust for `ApiOptions` parameter
  - Add `null` check for `ApiOptions`

- [ ] Add pagination Integration tests
  - `ReturnsCorrectCountOfXWithoutStart`
  - `ReturnsCorrectCountOfXWithStart`
  - `ReturnsDistinctXBasedOnStartPage`

### PR DESCRIPTION
**Pull request by gdziadkiewicz**
Fixes #1664 

- [x] Remove convention test exclusion

- [x] Add `ApiOptions` overload
  - make existing methods call new overload with `ApiOptions.None`

- [x] Update unit tests
  - Adjust for `ApiOptions` parameter
  - Add `null`/`""` checks for `ApiOptions` overloads

- [x] Add pagination Integration tests
  - `ReturnsCorrectCountOfXWithoutStart`
  - `ReturnsCorrectCountOfXWithStart`
  - `ReturnsDistinctXBasedOnStartPage`

- [x] Decide what to do with `GetAllForSubNamespace` endpoint

### CODE REVIEW

File: Octokit.Tests/Clients/ReferencesClientTests.cs [hunk 1]
```diff
@@ -106,8 +110,10 @@ public class TheGetAllForSubNamespaceMethod
                 await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace(null, "name", "heads"));
                 await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace("owner", null, "heads"));
                 await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace("owner", "name", null));
+                await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace("owner", "name", "heads", null));
 
                 await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace(1, null));
+                await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace(1, "subnamespace", null));
 
                 await Assert.ThrowsAsync<ArgumentException>(() => client.GetAllForSubNamespace("", "name", "heads"));
                 await Assert.ThrowsAsync<ArgumentException>(() => client.GetAllForSubNamespace("owner", "", "heads"));
```
Comment (ryangribble) [inline Octokit.Tests/Clients/ReferencesClientTests.cs]: we also need empty string test cases for the 4 parameter overload (and below for the 3 parameter one `1, "", options)`)

**Comment by gdziadkiewicz** [comment]
@ryangribble I will work on it today. I was low on spare time in past few days, sorry for the delay.

**Comment by ryangribble** [comment]
No apology necessary,  we all do this in our spare time when life permits 😝 

just wanted to make sure you weren't stuck

**Comment by gdziadkiewicz** [comment]
@ryangribble
I have added the missing unit tests, added and ran integration tests, but there is a problem - `GetAllForSubNamespace` endpoint ignores requested pagination. Do you have any idea how to handle that?

**Comment by ryangribble** [comment]
I've raised a ticket with GitHub support about the API pagination not being honoured for the subnamespace call

Repro:

✅  https://api.github.com/repos/octokit/octokit.net/git/refs?per_page=1

💀 https://api.github.com/repos/octokit/octokit.net/git/refs/tags?per_page=1

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs [hunk 2]
```diff
@@ -63,6 +63,61 @@ public ReferencesClientTests()
...
+        {
+            PageSize = 5,
+            PageCount = 1,
+            StartPage = 2
+        };
+
+        var firstRefsPage = await _fixture.GetAll("octokit", "octokit.net", startOptions);
+        var secondRefsPage = await _fixture.GetAll("octokit", "octokit.net", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: The nested loops are a bit clunky, is there a more succinct way to express this?  

eg `Assert.False(firstRefsPage.Any(x => secondRefsPage.Contains(x)));`

Comment (gdziadkiewicz) [reply Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: Changed it.

**Comment by ryangribble** [comment]
To get the travis build passing, pull in latest master to your branch :+1:

In terms of the pagination issue up stream, GitHub support confirmed it's a problem although it looks like it still isnt fixed.  I think it should be fine to leave it implemented in octokit as is, since hopefully they will fix it eventually 😁

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs [hunk 3]
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
...
+        {
+            PageSize = 5,
+            PageCount = 1,
+            StartPage = 2
+        };
+
+        var firstRefsPage = await _fixture.GetAll(7528679, startOptions);
+        var secondRefsPage = await _fixture.GetAll(7528679, skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: another assertion to refactor here

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs [hunk 4]
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
...
+        {
+            PageSize = 5,
+            PageCount = 1,
+            StartPage = 2
+        };
+
+        var firstRefsPage = await _fixture.GetAllForSubNamespace("octokit", "octokit.net", "heads", startOptions);
+        var secondRefsPage = await _fixture.GetAllForSubNamespace("octokit", "octokit.net", "heads", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: and here

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs [hunk 5]
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
...
+        {
+            PageSize = 5,
+            PageCount = 1,
+            StartPage = 2
+        };
+
+        var firstRefsPage = await _fixture.GetAllForSubNamespace(7528679, "heads", startOptions);
+        var secondRefsPage = await _fixture.GetAllForSubNamespace(7528679, "heads", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: and here

Comment (gdziadkiewicz) [reply Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: Done.

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs (hunk 4 above)
Comment (gdziadkiewicz) [reply Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: Done.

File: Octokit.Tests.Integration/Clients/ReferencesClientTests.cs (hunk 3 above)
Comment (gdziadkiewicz) [reply Octokit.Tests.Integration/Clients/ReferencesClientTests.cs]: Done.

**Comment by ryangribble** [comment]
pull_request: Add pagination support to `ReferencesClient`

**Comment by ryangribble** [comment]
Thanks @gdziadkiewicz !!!

**Pull request merged**